### PR TITLE
Remove duplicate code between `intersects_ray` and `intersects_segment`

### PR DESCRIPTION
**Pull request by Flarkk**
In `AABB`, `Plane`, `Geometry3D` _[Edit : and `TriangleMesh`]_ functions `ray_intersects_*` and `segment_intersects_*` involve almost identical code. In some cases both flavors even diverged over time which complexifies maintenance.

This PR :
- extends `ray_intersects_*` with an additional parameter `distance` that clamps the raycast in the same fashion as `segment_intersects_*`. `distance` is defaulted to `0.0` which enables infinite raycasting 
- makes `segment_intersects_*` wrap calls to `ray_intersects_*`. It avoids duplication and opens the way to eventually deprecating `intersects_segment_*` in the future, as they suffer from numerical precision issues (see related issue #100475)
- adds the `Geometry3D::ray_intersects_*` counterparts to `segment_intersects_sphere`, `_cylinder` and `_convex`
- enhances tests of `AABB` and `Geometry3D` to validate the above changes

### CODE REVIEW

File: tests/core/math/test_aabb.h [hunk 1]
```diff
@@ -186,24 +186,52 @@ TEST_CASE("[AABB] Intersection") {
 			!aabb_big.intersects_plane(Plane(Vector3(0, 1, 0), 200)),
 			"intersects_plane() should return the expected result.");
 
+	Vector3 result, normal;
```
Comment (AThousandShips) [inline tests/core/math/test_aabb.h:189]:
```suggestion
	Vector3 result;
	Vector3 normal;
```
Prefer one declaration per line

File: tests/core/math/test_geometry_3d.h [hunk 2]
```diff
@@ -147,34 +147,73 @@ TEST_CASE("[Geometry3D] Is Point in Projected Triangle") {
...
 	CHECK(Geometry3D::segment_intersects_convex(Vector3(10, 10, 10), Vector3(0, 0, 0), &box_planes[0], box_planes.size(), &result, &normal) == true);
 	CHECK(Geometry3D::segment_intersects_convex(Vector3(10, 10, 10), Vector3(5, 5, 5), &box_planes[0], box_planes.size(), &result, &normal) == true);
 	CHECK(Geometry3D::segment_intersects_convex(Vector3(10, 10, 10), Vector3(6, 5, 5), &box_planes[0], box_planes.size(), &result, &normal) == false);
 }
 
+TEST_CASE("[Geometry3D] Ray Intersects Convex") {
+	constexpr real_t sqrt2 = 1.41421356237309504880;
+	constexpr real_t sqrt3 = 1.7320508075688772935274463415059;
+	Vector<Plane> box_planes = Geometry3D::build_box_planes(Vector3(5, 5, 5));
+	Vector3 result, normal;
```
Comment (AThousandShips) [inline tests/core/math/test_geometry_3d.h:169]:
```suggestion
	Vector3 result;
	Vector3 normal;
```

File: tests/core/math/test_geometry_3d.h [hunk 3]
```diff
@@ -147,34 +147,73 @@ TEST_CASE("[Geometry3D] Is Point in Projected Triangle") {
...
+	CHECK(result.is_equal_approx(Vector3(2.5, 2.5, 2.5)));
+	CHECK(normal.is_equal_approx(Vector3(0, 0, 1)));
 	CHECK(Geometry3D::segment_intersects_cylinder(Vector3(10, 10, 10), Vector3(6, 6, 6), 5, 5, &result, &normal) == false);
 }
 
-TEST_CASE("[Geometry3D] Segment Intersects Cylinder") {
+TEST_CASE("[Geometry3D] Ray Intersects Cylinder") {
+	constexpr real_t sqrt2 = 1.41421356237309504880;
+	constexpr real_t sqrt3 = 1.7320508075688772935274463415059;
+	Vector3 result, normal;
```
Comment (AThousandShips) [inline tests/core/math/test_geometry_3d.h:187]:
```suggestion
	Vector3 result;
	Vector3 normal;
```

File: tests/core/math/test_geometry_3d.h [hunk 4]
```diff
@@ -147,34 +147,73 @@ TEST_CASE("[Geometry3D] Is Point in Projected Triangle") {
...
 	Vector3 result, normal;
 	CHECK(Geometry3D::segment_intersects_sphere(Vector3(10, 10, 10), Vector3(0, 0, 0), Vector3(0, 0, 0), 5, &result, &normal) == true);
 	CHECK(Geometry3D::segment_intersects_sphere(Vector3(10, 10, 10), Vector3(0, 0, 2.5), Vector3(0, 0, 0), 5, &result, &normal) == true);
 	CHECK(Geometry3D::segment_intersects_sphere(Vector3(10, 10, 10), Vector3(5, 5, 5), Vector3(0, 0, 0), 5, &result, &normal) == false);
 }
 
+TEST_CASE("[Geometry3D] Ray Intersects Sphere") {
+	constexpr real_t sqrt2 = 1.41421356237309504880;
+	constexpr real_t sqrt3 = 1.7320508075688772935274463415059;
+	Vector3 result, normal;
```
Comment (AThousandShips) [inline tests/core/math/test_geometry_3d.h:209]:
```suggestion
	Vector3 result;
	Vector3 normal;
```

File: tests/core/math/test_aabb.h [hunk 5]
```diff
@@ -186,24 +186,52 @@ TEST_CASE("[AABB] Intersection") {
...
+	result = Vector3();
+	normal = Vector3();
```
Comment (AThousandShips) [inline tests/core/math/test_aabb.h:191]:
```suggestion
```
Unnecessary first time

File: core/math/plane.h [hunk 6]
```diff
@@ -57,7 +57,7 @@ struct [[nodiscard]] Plane {
 	/* intersections */
 
 	bool intersect_3(const Plane &p_plane1, const Plane &p_plane2, Vector3 *r_result = nullptr) const;
-	bool intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *p_intersection) const;
+	bool intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *p_intersection, real_t max_distance = 0) const;
```
Comment (AThousandShips) [inline core/math/plane.h:60]:
```suggestion
	bool intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *p_intersection, real_t p_max_distance = 0) const;
```

File: core/math/plane.cpp [hunk 7]
```diff
@@ -94,7 +94,7 @@ bool Plane::intersect_3(const Plane &p_plane1, const Plane &p_plane2, Vector3 *r
 	return true;
 }
 
-bool Plane::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *p_intersection) const {
+bool Plane::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *p_intersection, real_t max_distance) const {
```
Comment (AThousandShips) [inline core/math/plane.cpp:97]:
```suggestion
bool Plane::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *p_intersection, real_t p_max_distance) const {
```

File: core/math/geometry_3d.h [hunk 8]
```diff
@@ -313,20 +275,30 @@ class Geometry3D {
...
+			*r_norm = p_planes[min_index].normal;
 		}
 
 		return true;
 	}
 
+	static inline bool segment_intersects_convex(const Vector3 &p_from, const Vector3 &p_to, const Plane *p_planes, int p_plane_count, Vector3 *r_res, Vector3 *r_norm) {
+		Vector3 segment = p_to - p_from;
+		real_t length = segment.length();
+		if (unlikely(Math::is_equal_approx(length, 0))) {
```
Comment (AThousandShips) [inline core/math/geometry_3d.h:295]:
```suggestion
		if (unlikely(Math::is_zero_approx(length))) {
```

File: core/math/geometry_3d.h [hunk 9]
```diff
@@ -274,17 +233,20 @@ class Geometry3D {
...
 			return false;
 		}
+		Vector3 ray = segment / length;
+		return ray_intersects_cylinder(p_from, ray, p_height, p_radius, r_res, r_norm, p_cylinder_axis, length);
+	}
 
-		Vector3 dir = rel / rel_l;
+	static bool ray_intersects_convex(const Vector3 &p_from, const Vector3 &p_dir, const Plane *p_planes, int p_plane_count, Vector3 *r_res, Vector3 *r_norm, real_t p_dist = 0.0) {
+		real_t min = -1e20, max = 1e20;
```
Comment (AThousandShips) [inline core/math/geometry_3d.h:247]:
```suggestion
		real_t min = -1e20;
		real_t max = 1e20;
```

File: core/math/geometry_3d.h [hunk 10]
```diff
@@ -72,65 +72,28 @@ class Geometry3D {
 		// the intersection point is on the line.
 		real_t t = f * e2.dot(q);
 
-		if (t > 0.00001f) { // ray intersection
+		if (t > (real_t)CMP_EPSILON && (p_dist <= 0 || t <= p_dist)) { // ray intersection and check if distance is within boundary if provided
```
Comment (AThousandShips) [inline core/math/geometry_3d.h:75]:
```suggestion
		if (t > (real_t)CMP_EPSILON && (p_dist <= 0 || t <= p_dist)) { // Ray intersection and check if distance is within boundary if provided.
```

File: core/math/aabb.h [hunk 11]
```diff
@@ -73,12 +73,27 @@ struct [[nodiscard]] AABB {
 	AABB intersection(const AABB &p_aabb) const; ///get box where two intersect, empty if no intersection occurs
 	_FORCE_INLINE_ bool smits_intersect_ray(const Vector3 &p_from, const Vector3 &p_dir, real_t p_t0, real_t p_t1) const;
 
-	bool intersects_segment(const Vector3 &p_from, const Vector3 &p_to, Vector3 *r_intersection_point = nullptr, Vector3 *r_normal = nullptr) const;
+	bool intersects_segment(const Vector3 &p_from, const Vector3 &p_to, Vector3 *r_intersection_point = nullptr, Vector3 *r_normal = nullptr) const {
+		Vector3 segment = p_to - p_from;
+		real_t length = segment.length();
+		Vector3 ray = Math::is_equal_approx(length, 0) ? Vector3() : segment / length;
```
Comment (AThousandShips) [inline core/math/aabb.h:79]:
```suggestion
		Vector3 ray = Math::is_zero_approx(length) ? Vector3() : segment / length;
```

File: core/math/aabb.h [hunk 12]
```diff
@@ -73,12 +73,27 @@ struct [[nodiscard]] AABB {
...
+		bool inside;
```
Comment (AThousandShips) [inline core/math/aabb.h:80]:
```suggestion
		bool inside = false;
```

File: core/math/geometry_3d.h [hunk 13]
```diff
@@ -202,55 +169,47 @@ class Geometry3D {
...
-			if (seg_from < seg_to) {
-				if (seg_from > box_end || seg_to < box_begin) {
+			if (dir_2d[i] == 0) {
+				if ((from_2d[i] < -size_2d[i]) || (from_2d[i] > size_2d[i])) {
 					return false;
 				}
-				real_t length = seg_to - seg_from;
-				cmin = (seg_from < box_begin) ? ((box_begin - seg_from) / length) : 0;
-				cmax = (seg_to > box_end) ? ((box_end - seg_from) / length) : 1;
+			} else { // ray not parallel to bounds in this direction
```
Comment (AThousandShips) [inline core/math/geometry_3d.h:188]:
```suggestion
			} else { // Ray not parallel to bounds in this direction.
```

File: core/math/geometry_3d.h [hunk 14]
```diff
@@ -274,17 +233,20 @@ class Geometry3D {
...
-		real_t min = -1e20, max = 1e20;
-
-		Vector3 rel = p_to - p_from;
-		real_t rel_l = rel.length();
-
-		if (rel_l < (real_t)CMP_EPSILON) {
+	static inline bool segment_intersects_cylinder(const Vector3 &p_from, const Vector3 &p_to, real_t p_height, real_t p_radius, Vector3 *r_res = nullptr, Vector3 *r_norm = nullptr, int p_cylinder_axis = 2) {
+		Vector3 segment = p_to - p_from;
+		real_t length = segment.length();
+		if (unlikely(Math::is_equal_approx(length, 0))) {
```
Comment (AThousandShips) [inline core/math/geometry_3d.h:239]:
```suggestion
		if (unlikely(Math::is_zero_approx(length))) {
```

File: core/math/geometry_3d.h [hunk 15]
```diff
@@ -72,65 +72,28 @@ class Geometry3D {
...
-
-		if (t > (real_t)CMP_EPSILON && t <= 1.0f) { // Ray intersection.
-			if (r_res) {
-				*r_res = p_from + rel * t;
-			}
-			return true;
-		} else { // This means that there is a line intersection but not a ray intersection.
+		Vector3 segment = p_to - p_from;
+		real_t length = segment.length();
+		if (unlikely(Math::is_equal_approx(length, 0))) {
```
Comment (AThousandShips) [inline core/math/geometry_3d.h:87]:
```suggestion
		if (unlikely(Math::is_zero_approx(length))) {
```

File: core/math/geometry_3d.h [hunk 16]
```diff
@@ -164,20 +127,24 @@ class Geometry3D {
...
 
-	static inline bool segment_intersects_cylinder(const Vector3 &p_from, const Vector3 &p_to, real_t p_height, real_t p_radius, Vector3 *r_res = nullptr, Vector3 *r_norm = nullptr, int p_cylinder_axis = 2) {
-		Vector3 rel = (p_to - p_from);
-		real_t rel_l = rel.length();
-		if (rel_l < (real_t)CMP_EPSILON) {
-			return false; // Both points are the same.
+	static inline bool segment_intersects_sphere(const Vector3 &p_from, const Vector3 &p_to, const Vector3 &p_sphere_pos, real_t p_sphere_radius, Vector3 *r_res = nullptr, Vector3 *r_norm = nullptr) {
+		Vector3 segment = p_to - p_from;
+		real_t length = segment.length();
+		if (unlikely(Math::is_equal_approx(length, 0))) {
```
Comment (AThousandShips) [inline core/math/geometry_3d.h:133]:
```suggestion
		if (unlikely(Math::is_zero_approx(length))) {
```

File: tests/core/math/test_geometry_3d.h [hunk 17]
```diff
@@ -147,34 +147,73 @@ TEST_CASE("[Geometry3D] Is Point in Projected Triangle") {
...
 
+TEST_CASE("[Geometry3D] Ray Intersects Convex") {
+	constexpr real_t sqrt2 = 1.41421356237309504880;
+	constexpr real_t sqrt3 = 1.7320508075688772935274463415059;
+	Vector<Plane> box_planes = Geometry3D::build_box_planes(Vector3(5, 5, 5));
+	Vector3 result, normal;
+	CHECK(Geometry3D::ray_intersects_convex(Vector3(10, 10, 10), Vector3(-1.0 / sqrt3, -1.0 / sqrt3, -1.0 / sqrt3), &box_planes[0], box_planes.size(), &result, &normal) == true);
+	CHECK(Geometry3D::ray_intersects_convex(Vector3(0, 0, 1), Vector3(1.0 / sqrt3, 1.0 / sqrt3, -1.0 / sqrt3), &box_planes[0], box_planes.size(), &result, &normal) == false);
+	CHECK(Geometry3D::ray_intersects_convex(Vector3(10, 10, 10), Vector3(-1.0 / sqrt3, -1.0 / sqrt3, -1.0 / sqrt3), &box_planes[0], box_planes.size(), &result, &normal, 2.0) == false);
+	CHECK(Geometry3D::ray_intersects_convex(Vector3(7, 7, 1), Vector3(-1.0 / sqrt2, -1.0 / sqrt2, 0), &box_planes[0], box_planes.size(), &result, &normal) == true);
```
Comment (AThousandShips) [inline tests/core/math/test_geometry_3d.h:173]:
```suggestion
	CHECK(Geometry3D::ray_intersects_convex(Vector3(10, 10, 10), Vector3(-1.0 / sqrt3, -1.0 / sqrt3, -1.0 / sqrt3), &box_planes[0], box_planes.size(), &result, &normal));
	CHECK_FALSE(Geometry3D::ray_intersects_convex(Vector3(0, 0, 1), Vector3(1.0 / sqrt3, 1.0 / sqrt3, -1.0 / sqrt3), &box_planes[0], box_planes.size(), &result, &normal));
	CHECK_FALSE(Geometry3D::ray_intersects_convex(Vector3(10, 10, 10), Vector3(-1.0 / sqrt3, -1.0 / sqrt3, -1.0 / sqrt3), &box_planes[0], box_planes.size(), &result, &normal, 2.0));
	CHECK(Geometry3D::ray_intersects_convex(Vector3(7, 7, 1), Vector3(-1.0 / sqrt2, -1.0 / sqrt2, 0), &box_planes[0], box_planes.size(), &result, &normal));
```

File: tests/core/math/test_geometry_3d.h [hunk 18]
```diff
@@ -147,34 +147,73 @@ TEST_CASE("[Geometry3D] Is Point in Projected Triangle") {
...
+	CHECK(normal.is_equal_approx(Vector3(0, 0, 1)));
 	CHECK(Geometry3D::segment_intersects_cylinder(Vector3(10, 10, 10), Vector3(6, 6, 6), 5, 5, &result, &normal) == false);
 }
 
-TEST_CASE("[Geometry3D] Segment Intersects Cylinder") {
+TEST_CASE("[Geometry3D] Ray Intersects Cylinder") {
+	constexpr real_t sqrt2 = 1.41421356237309504880;
+	constexpr real_t sqrt3 = 1.7320508075688772935274463415059;
+	Vector3 result, normal;
+	CHECK(Geometry3D::ray_intersects_cylinder(Vector3(10, 10, 10), Vector3(-1.0 / sqrt3, -1.0 / sqrt3, -1.0 / sqrt3), 50, 5, &result, &normal) == true);
```
Comment (AThousandShips) [inline tests/core/math/test_geometry_3d.h:188]:
```suggestion
	CHECK(Geometry3D::ray_intersects_cylinder(Vector3(10, 10, 10), Vector3(-1.0 / sqrt3, -1.0 / sqrt3, -1.0 / sqrt3), 50, 5, &result, &normal));
```

File: tests/core/math/test_geometry_3d.h [hunk 19]
```diff
@@ -147,34 +147,73 @@ TEST_CASE("[Geometry3D] Is Point in Projected Triangle") {
...
+	CHECK(result.is_equal_approx(Vector3(5.0 * sqrt2 / 2.0, 5.0 * sqrt2 / 2.0, 5.0 * sqrt2 / 2.0)));
+	CHECK(normal.is_equal_approx(Vector3(1.0 / sqrt2, 1.0 / sqrt2, 0)));
+	CHECK(Geometry3D::ray_intersects_cylinder(Vector3(10, 10, 10), Vector3(-1.0 / sqrt3, -1.0 / sqrt3, -1.0 / sqrt3), 5, 5, &result, &normal, 2, 3.0) == false);
+	CHECK(Geometry3D::ray_intersects_cylinder(Vector3(1, 10, -1), Vector3(0, -1, 0), 5, 5, &result, &normal, 1) == true);
```
Comment (AThousandShips) [inline tests/core/math/test_geometry_3d.h:192]:
```suggestion
	CHECK_FALSE(Geometry3D::ray_intersects_cylinder(Vector3(10, 10, 10), Vector3(-1.0 / sqrt3, -1.0 / sqrt3, -1.0 / sqrt3), 5, 5, &result, &normal, 2, 3.0));
	CHECK(Geometry3D::ray_intersects_cylinder(Vector3(1, 10, -1), Vector3(0, -1, 0), 5, 5, &result, &normal, 1));
```

File: tests/core/math/test_geometry_3d.h [hunk 20]
```diff
@@ -147,34 +147,73 @@ TEST_CASE("[Geometry3D] Is Point in Projected Triangle") {
...
+	CHECK(result.is_equal_approx(Vector3(1, 2.5, -1)));
+	CHECK(normal.is_equal_approx(Vector3(0, 1, 0)));
+	CHECK(Geometry3D::ray_intersects_cylinder(Vector3(1, 10, -1), Vector3(0, -1, 0), 5, 5, &result, &normal, 2, 2.0) == false);
+	CHECK(Geometry3D::ray_intersects_cylinder(Vector3(10, 10, 10), Vector3(1.0 / sqrt2, 0, 1.0 / sqrt2), 5, 5, &result, &normal) == false);
```
Comment (AThousandShips) [inline tests/core/math/test_geometry_3d.h:196]:
```suggestion
	CHECK_FALSE(Geometry3D::ray_intersects_cylinder(Vector3(1, 10, -1), Vector3(0, -1, 0), 5, 5, &result, &normal, 2, 2.0));
	CHECK_FALSE(Geometry3D::ray_intersects_cylinder(Vector3(10, 10, 10), Vector3(1.0 / sqrt2, 0, 1.0 / sqrt2), 5, 5, &result, &normal));
```

File: tests/core/math/test_geometry_3d.h [hunk 21]
```diff
@@ -147,34 +147,73 @@ TEST_CASE("[Geometry3D] Is Point in Projected Triangle") {
...
 
+TEST_CASE("[Geometry3D] Ray Intersects Sphere") {
+	constexpr real_t sqrt2 = 1.41421356237309504880;
+	constexpr real_t sqrt3 = 1.7320508075688772935274463415059;
+	Vector3 result, normal;
+	CHECK(Geometry3D::ray_intersects_sphere(Vector3(-25, -25, -25), Vector3(1.0 / sqrt3, 1.0 / sqrt3, 1.0 / sqrt3), Vector3(0, -1, 0), 5, &result, &normal) == true);
+	CHECK(Geometry3D::ray_intersects_sphere(Vector3(-25, -25, -25), Vector3(1.0 / sqrt3, 1.0 / sqrt3, 1.0 / sqrt3), Vector3(0, -1, 0), 5, &result, &normal, 5.0) == false);
+	CHECK(Geometry3D::ray_intersects_sphere(Vector3(-25, -25, -25), Vector3(-1.0 / sqrt3, 1.0 / sqrt3, 1.0 / sqrt3), Vector3(0, 0, 1), 200, &result, &normal) == false);
+	CHECK(Geometry3D::ray_intersects_sphere(Vector3(-25, -25, -25), Vector3(1.0 / sqrt2, 1.0 / sqrt2, 0), Vector3(0, 0, -20), 8, &result, &normal) == true);
+	CHECK(Geometry3D::ray_intersects_sphere(Vector3(-25, -25, -25), Vector3(1.0 / sqrt2, 1.0 / sqrt2, 0), Vector3(0, 0, -20), 8, &result, &normal, 2.0) == false);
```
Comment (AThousandShips) [inline tests/core/math/test_geometry_3d.h:214]:
```suggestion
	CHECK(Geometry3D::ray_intersects_sphere(Vector3(-25, -25, -25), Vector3(1.0 / sqrt3, 1.0 / sqrt3, 1.0 / sqrt3), Vector3(0, -1, 0), 5, &result, &normal));
	CHECK_FALSE(Geometry3D::ray_intersects_sphere(Vector3(-25, -25, -25), Vector3(1.0 / sqrt3, 1.0 / sqrt3, 1.0 / sqrt3), Vector3(0, -1, 0), 5, &result, &normal, 5.0));
	CHECK_FALSE(Geometry3D::ray_intersects_sphere(Vector3(-25, -25, -25), Vector3(-1.0 / sqrt3, 1.0 / sqrt3, 1.0 / sqrt3), Vector3(0, 0, 1), 200, &result, &normal));
	CHECK(Geometry3D::ray_intersects_sphere(Vector3(-25, -25, -25), Vector3(1.0 / sqrt2, 1.0 / sqrt2, 0), Vector3(0, 0, -20), 8, &result, &normal));
	CHECK_FALSE(Geometry3D::ray_intersects_sphere(Vector3(-25, -25, -25), Vector3(1.0 / sqrt2, 1.0 / sqrt2, 0), Vector3(0, 0, -20), 8, &result, &normal, 2.0));
```

**Comment by Flarkk** [comment]
@lawnjelly here the old functions now wrap a call to the new ones as you suggested in #100475. I believe it's helpful here.

**Comment by Flarkk** [comment]
Resolved requested changes. Thanks @AThousandShips for your torough review.

File: core/math/aabb.cpp [hunk 22]
```diff
@@ -159,7 +159,7 @@ bool AABB::find_intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, bool
 				}
 				tmax = t2;
 			}
-			if (tmin > tmax) {
+			if (tmin > tmax || (p_max_distance > 0 && tmin > p_max_distance)) {
```
Comment (lawnjelly) [inline core/math/aabb.cpp:162]: Are we now paying the price for this check even when not used? (`p_max_distance` defaulted to 0?)

I can see the `p_max_distance > 0` should be optimized out but does the second term `tmin > p_max_distance`?

**Comment by lawnjelly** [comment]
> In some cases both flavors even diverged over time which complexifies maintenance.

Agree, but it does mean that any changes to behaviour could cause knock on bugs. I haven't looked in detail but are there any changes to behaviour here, or is it purely moving code around?

**Comment by Flarkk** [comment]
Added `TriangleMesh` to the list of simplifications.

File: core/math/aabb.cpp (hunk 22 above)
Comment (Flarkk) [reply core/math/aabb.cpp:162]: > Are we now paying the price for this check even when not used? (`p_max_distance` defaulted to 0?)
> 
> I can see the `p_max_distance > 0` should be optimized out but does the second term `tmin > p_max_distance`?

Yes we are. I believe nothing can be optimized out right now because the code is in the .cpp.
Will consider templating the function (therefore move it to the header file).

**Comment by Flarkk** [comment]
Pushed some small changes in relation to the below.

> Agree, but it does mean that any changes to behaviour could cause knock on bugs. I haven't looked in detail but are there any changes to behaviour here, or is it purely moving code around?

Generally speaking the intention is to preserve the exact original behavior.
Most changes just consist in wrapping calls and replacing all `to - from` by `direction` and `distance`, except some subtleties with edge cases.

TLDR :
- `AABB` : some differences in edge cases
- `Plane` : same behavior
- `Geometry3D` : same behavior except 1 difference with Sphere intersection
- `TriangleMesh` : same behavior

<details>
  <summary>Unfold details</summary>

### `AABB`
Original `segment` and `ray` intersection implementations are based on distinct algorithms.
With this PR, `segment` intersection now runs on `ray`'s code, which has identical behavior except on few edge cases :
 - **Ray starting point inside the AABB** : `ray` intersection returns the entry point _before_ the ray starting point when the latter is located _inside_ the box, where `segment` intersection returns the ray starting point itself. A consequence of it is `segment` returns unspecified normals for ray starting points inside the box.
:information_source: This PR preserves these differences in the `segment` version, except the normal is now set to `Vector3()` when the starting point is inside the box
- **Zero-length segments** : when ray start and end are strictly identical `segment` version behaves like a point intersection (true if inside, false otherwise).
:heavy_check_mark: This PR preserves this behavior by falling back to `has_point()` when necessary.
- **Exact intersection coordinates** : `ray` version ensures returned intersection coordinates *exactly* lie on the box edge to compensate for numerical precision roundings, when `segment` doesn't.
:information_source: This PR harmonizes the behavior and applies this correction in all cases
- **AABBs larger than 1e20** : original `ray` version have indefined results with such very large boxes whereas `segment` manages them correctly.
:heavy_check_mark: This PR fixes `ray` version behavior such that very large boxes are managed correctly in all cases
### `Plane`
No edge case. Strictly indentical implementation.
### `Geometry3D`
#### Triangle intersection
No edge case. Strictly indentical implementation.
#### Sphere intersection
- **Zero-length segments** : original `segment` implementation consider start and end points identical and return `false` below `EPSILON`.
:information_source: This PR changes this for an exact 0 length check to keep it harmonized with all other intersection functions.
#### Cylinder and Convex intersection
A subpart of the Cylinder's algorithm has been rewritten to get rid of `(to - from)`, however the new behavior is strictly identical.
- **Objects larger than 1e20** : original `segment` implementations of both Cylinder and Convex have indefined results with very large objects.
:heavy_check_mark: This PR fixes it to support any size
### `TriangleMesh`
- **Mesh with vertices farther than 1e10** : both original `ray` and `segment` implementations have indefined results with such very large meshes.
:heavy_check_mark: This PR fixes it to support any mesh size

</details>

But overall I agree this kind of change can have unintended side effects if not toroughly tested.
Unit tests are good but usually not enough.

I'd like to set up a demo project or something visual to allow stress-testing these changes.
It's not that simple because it's core classes and they're not directly accessible from scripts.
I didn't think through it yet but if you have any idea, let me know.

**Comment by Flarkk** [comment]
Also, I realized that defaulting `distance` to `0.0` to signify infinite raycasting doesn't allow testing against a point (i.e. a zero-length segment). It's yet another edge case, but the `from` / `to` versions do support it.

I'm not fully clear on whether it's a real problem. Any views welcome.

Alternatives could be :
- to default to `-1.0` instead
- to keep it as is but add `is_point_inside()` functions in all classes to allow testing against points
- in all cases this is better documented : what's the standard practice (except commenting on the code) to document internal classes ?

File: core/math/aabb.cpp (hunk 22 above)
Comment (lawnjelly) [reply core/math/aabb.cpp:162]: Afaik compiler can optimize out functions in the cpp if called from the same translation unit (i.e. it has the opportunity to inline different versions of called functions), putting stuff in the header allows stuff to be optimized from a _different_ translation unit. (Link time optimization can theoretically also remove this limit but may not be as smart as in the same translation unit.)

Afaik usually this should mean wrapper functions in the same header / cpp will aggressively optimize.

Always worth checking the disassembly to see what is going on, or e.g. profile release version and watch the hot spots.

**Comment by Flarkk** [comment]
Rebased
